File: 240109-search-sort/2-15-merge-sort.py
def merge_sort(arr):
    if len(arr) > 1:
        mean = len(arr)//2
        left_arr = arr[:mean]
        right_arr = arr[mean:]
        merge_sort(left_arr)
        merge_sort(right_arr)
 
        i = j = k = 0
        while i < len(left_arr) and j < len(right_arr):
            if left_arr[i] < right_arr[j]:
                arr[k] = left_arr[i]
                i += 1
            else:
                arr[k] = right_arr[j]
                j += 1
            k += 1
        while i < len(left_arr):
            arr[k] = left_arr[i]
            i += 1
            k += 1
 
        while j < len(right_arr):
            arr[k] = right_arr[j]
            j += 1
            k += 1    
 
library_num = [124,235,456,123,756,476,285,998,379,108]
print("Initial array:", library_num)
merge_sort(library_num)
print ("Sorted array:", library_num)

File: 240109-search-sort/2-15-merge-sort.c
#include <stdio.h>
 
 void merge(int items[], int left, int mean, int right) {
    int leftSize = mean - left + 1;
    int rightSize = right - mean;
    int leftArr[leftSize], rightArr[rightSize];
    for (int i = 0; i < leftSize; i++)
        leftArr[i] = items[left + i];
    for (int j = 0; j < rightSize; j++)
        rightArr[j] = items[mean + 1 + j];
 
  // Maintain current index of sub-arrays and main array
  int i = 0;
  int j = 0;
  int k = left;
 
  // Until we reach the end of either leftArr or rightArr, pick larger among
  // elements leftArr and rightArr and place them in the correct position at items[left..right]
  while(i < leftSize && j < rightSize) {
    if (leftArr[i] <= rightArr[j]) {
        items[k] = leftArr[i];
        i++;
        } else {
            items[k] = rightArr[j];
            j++;
            }
    k++;
  }
  while (i < leftSize) {
    items[k] = leftArr[i];
    i++;
    k++;
    }
    while (j < rightSize) {
        items[k] = rightArr[j];
        j++;
        k++;
        }
}
 
// Divide the array into two subarrays, sort them and merge them
void mergeSort(int arr[], int left, int right) {
    if (left < right) {
        int mean  = left + (right - left) / 2;
        mergeSort(arr, left, mean );
        mergeSort(arr, mean  + 1, right);
        merge(arr, left, mean , right);
        }
}
 
void showArray(int arr[], int size) 
{ 
    int i; 
    for (i=0; i < size; i++) 
        printf("%d ", arr[i]); 
    printf("\n"); 
}
 
int main()
{ 
    int libraryNum[] = {124,235,456,123,756,476,285,998,379,108}; 
    int n = sizeof(libraryNum)/sizeof(libraryNum[0]); 
 
    printf("Initial array: \n"); 
    showArray(libraryNum, n); 
    mergeSort(libraryNum, 0, n-1); 
    printf("Sorted array: \n"); 
    showArray(libraryNum, n); 
    return 0; 
}